src/listpack.c

```diff
@@ -623,17 +650,104 @@ unsigned char *lpGet(unsigned char *p, int64_t *count, unsigned char *intbuf) {
+    unsigned char *value;
+    int64_t ll, vll;
+    uint64_t entry_size = 123456789; /* initialized to avoid warning. */
+    uint32_t lp_bytes = lpBytes(lp);
+
+    assert(p);
+    while (p) {
+        if (skipcnt == 0) {
+            value = lpGetWithSize(p, &ll, NULL, &entry_size);
+            if (value) {
+                if (slen == ll && memcmp(value, s, slen) == 0) {
+                    return p;
+                }
+            } else {
+                /* Find out if the searched field can be encoded. Note that
+                 * we do it only the first time, once done vencoding is set
+                 * to non-zero and vll is set to the integer value. */
+                if (vencoding == 0) {
+                    /* If the entry can be encoded as integer we set it to
+                     * 1, else set it to UCHAR_MAX, so that we don't retry
+                     * again the next time. */
+                    if (slen >= 32 || slen == 0 || !lpStringToInt64((const char*)s, slen, &vll)) {
+                        vencoding = UCHAR_MAX;
+                    } else {
+                        vencoding = 1;
+                    }
+                }
+
+                /* Compare current entry with specified entry, do it only
+                 * if vencoding != UCHAR_MAX because if there is no encoding
+                 * possible for the field it can't be a valid integer. */
+                if (vencoding != UCHAR_MAX && ll == vll) {
+                    return p;
+                }
+            }
+
+            /* Reset skip count */
+            skipcnt = skip;
+            p += entry_size;
+        } else {
+            /* Skip entry */
+            skipcnt--;
+
+            /* Move to next entry, avoid use `lpNext` due to `ASSERT_INTEGRITY` in
+            * `lpNext` will call `lpBytes`, will cause performance degradation */
+            p = lpSkip(p);
+        }
+
+        assert(p >= lp + LP_HDR_SIZE && p < lp + lp_bytes);
+        if (p[0] == LP_EOF) break;
+    }
+
+    return NULL;
+}
+
 /* Insert, delete or replace the specified element 'ele' of length 'len' at
  * the specified position 'p', with 'p' being a listpack element pointer
  * obtained with lpFirst(), lpLast(), lpNext(), lpPrev() or lpSeek().
  *
  * The element is inserted before, after, or replaces the element pointed
  * by 'p' depending on the 'where' argument, that can be LP_BEFORE, LP_AFTER
  * or LP_REPLACE.
- *
- * If 'ele' is set to NULL, the function removes the element pointed by 'p'
- * instead of inserting one.
- *
+ * 
+ * If 'enctype'==-1, 'ele' is NULL and the function removes the element pointed by 'p'
+ *                   instead of inserting one.
+ * If 'enctype'==LP_ENCODING_INT, the function insert or replace with a 64 bit integer,
+ *                   which is stored in the 'intenc' buffer.
+ * If 'enctype'==LP_ENCODING_STR, the function insert or replace with a string,
+ *                   which is stored in the 'ele' buffer.
+ * 
```

Comment:
i think maybe we can do without the `enctype` argument
```suggestion
 * If both 'elestr' and `eleint` are NULL and the function removes the element pointed by 'p'
 *                   instead of inserting one.
 * If `eleint` is non-NULL, the function insert or replace with a 64 bit integer,
 *                   which is stored in the 'eleint' buffer.
 * If 'elestr` is non-NULL, the function insert or replace with a string,
 *                   which is stored in the 'elestr' buffer.
 * 
```

internally the function can maybe still use enctype like it used to use.. just the interface may be a bit cleaner>

WDYT?

p.s. we're also missing some documentation of `enclen`, right?

obviously if we take that, we need to update the other references to `ele` in the code and comments.

Reply:
ohh, I originally wanted to design it as ```lpInsert(unsigned char *lp, unsigned char *ele, uint32_t size, uint64_t lval, ...)```,
But I found it impossible to tell if it is deletion.
The way you said it is the right way.

Reply:
Fixed.